### PR TITLE
Add a justfile with targets from doit

### PR DESCRIPTION
**Pull request by FollowTheProcess**
Related to discussion item: #128 

This PR adds a `justfile` to the root of the project with targets taken from the `dodo.py` configuration file.

I've run all the recipes locally and they seem to be working fine 🙂

There's a few differences between this and the `dodo.py` implementation:

- Because `join` currently only accepts 2 arguments (pending next release) I've had to declare a few more variables to get proper file paths
- I've changed the default task to just print the task list while `just` is still new to the project (happy to change to reflect default tasks in `dodo.py`, thought I'd keep it safe for now!)
- I tried to get the Graphviz target running in a for loop similar to `dodo.py` so you could just pass an array of formats but couldn't get `just`'s variable substitution and shell variable expansion playing together nicely so I've just got it as two lines in the task for now

I appreciate I didn't give you chance to respond in the discussion before throwing this PR at you but I was bored and this seemed like fun!

Please feel free to reject if this is no longer wanted 🙂 

Thanks!

### CODE REVIEW

File: justfile [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    import datetime
+    import pathlib
+    import re
+
+    with open("{{ CARGO_TOML }}", "r", encoding="utf-8") as file:
+        cargo_lines = file.readlines()
+
+    for line in cargo_lines:
+        version_match = re.match(r'version\s*=\s*"(?P<version>[\d.]+)"', line)
+        if version_match:
```
Comment (brettcannon) [inline justfile:53]:
```suggestion
        if version_match := VERSION_REGEX.match(line):
```

File: justfile [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    pandoc {{ MAN_MD }} --standalone -t man -o {{ MAN_FILE }}
+
+# Build the man page
+man: _man-md
+    #!/usr/bin/env python3
+
+    import datetime
+    import pathlib
+    import re
+
```
Comment (brettcannon) [inline justfile:47]:
```suggestion
    VERSION_REGEX = re.compile(r'version\s*=\s*"(?P<version>[\d.]+)"')

```

File: justfile [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        "CURRENT_DATE", datetime.date.today().isoformat()
+    )
+
+    with open("{{ MAN_FILE }}", "w", encoding="utf-8") as file:
+        file.write(new_man_text)
+
+# Build the control flow graphviz diagram
+dot:
+    dot -T "svg" -o {{ DOT_SVG }} {{ DOT_FILE }}
+    dot -T "png" -o {{ DOT_PNG }} {{ DOT_FILE }}
```
Comment (brettcannon) [inline justfile:73]: Do you think it's worth breaking this up into separate rules? My thinking is this way you can choose to build just one of them while you iterate on the control flow diagram. Plus it should help parallelize the build overall.

```suggestion

# Build the control flow diagram as an SVG
dot_svg:
    dot -T "svg" -o {{ DOT_SVG }} {{ DOT_FILE }}
    
# Build the control flow diagram as a PNG
dot_png:
    dot -T "png" -o {{ DOT_PNG }} {{ DOT_FILE }}
    
# Build the control flow diagram
dot: dot_svg dot_png
```

**Review by brettcannon — CHANGES_REQUESTED** [review]
Looks great! I have a couple of suggestions, but I'm not sure if they make sense since this is my first justfile. 😅

File: justfile [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+DOCS := join(ROOT, "docs")
+MAN_DIR := join(DOCS, "man-page")
+MAN_MD := join(MAN_DIR, "py.1.md")
+MAN_FILE := join(MAN_DIR, "py.1")
+CARGO_TOML := join(ROOT, "Cargo.toml")
+DOT_DIR := join(DOCS, "control-flow")
+DOT_FILE := join(DOT_DIR, "control_flow.dot")
+DOT_STEM := file_stem(DOT_FILE)
+DOT_SVG := join(DOT_DIR, DOT_STEM) + ".svg"
+DOT_PNG := join(DOT_DIR, DOT_STEM) + ".png"
```
[the remark this answers is not in the document]
Comment (FollowTheProcess) [reply justfile:15]: Nice! much cleaner

File: justfile (hunk 1 above)
Comment (FollowTheProcess) [reply justfile:53]: Do you think it's worth explaining somewhere that a contributor would need python >=3.8 (I think!) in order to run this recipe or is it mainly just for CI where you control the version entirely?

File: justfile (hunk 3 above)
Comment (FollowTheProcess) [reply justfile:73]: Yeah happy to do this 🙂

**Comment by FollowTheProcess** [comment]
@brettcannon Think I got all your suggestions in, let me know if I missed any. My only question was re the use of the walrus in the python script around maybe explaining somewhere that a contributor would nee python >=3.8 to run this recipe?

File: justfile (hunk 1 above)
Comment (brettcannon) [reply justfile]: The failure for folks not running on Python 3.8 should be obvious. And if even Debian ships with 3.8 then it's reasonable to accept folks will have it installed.

**Comment by brettcannon** [comment]
Just realized that `.gitignore` also now has a dead entry (found via https://github.com/brettcannon/python-launcher/search?q=doit).

File: justfile [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    man_text_with_version = man_text.replace("LAUNCHER_VERSION", version)
+    new_man_text = man_text_with_version.replace(
+        "CURRENT_DATE", datetime.date.today().isoformat()
+    )
+
+    with open("{{ MAN_FILE }}", "w", encoding="utf-8") as file:
+        file.write(new_man_text)
+
+# Build the control flow diagram as an SVG
```
Comment (brettcannon) [inline justfile:70]:
```suggestion
# Build the control flow diagram as a SVG
```

File: justfile [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+# Run linting on source files
+lint:
+    cargo fmt --quiet --all -- --check
+    cargo clippy --quiet --all-targets --all-features -- -D warnings
+
+# Install from source
+install:
+    cargo install --quiet --path .
+
+# Convert the markdown man to man file format (hidden)
```
Comment (brettcannon) [inline justfile:35]:
```suggestion
# Convert the markdown-formatted man page to the man file format (hidden)
```

File: justfile (hunk 1 above)
Comment (FollowTheProcess) [reply justfile]: Cool sounds good to me 🙂

**Comment by FollowTheProcess** [comment]
> Just realized that `.gitignore` also now has a dead entry (found via https://github.com/brettcannon/python-launcher/search?q=doit).

Yeah I left that in because I didn't know if you wanted to completely remove `doit`? Happy to take it out? Would you want `dodo.py` removing as part of this too?

I'll update the `Dockerfile` too 👍🏻

**Comment by brettcannon** [comment]
> Yeah I left that in because I didn't know if you wanted to completely remove doit? Happy to take it out? Would you want dodo.py removing as part of this too?

Yep, this PR is removing all use of doit from the project.

> I'll update the Dockerfile too 👍🏻

Thanks! And I'm not Docker expert, so if there's anything missing please let me know.

> As far as I can see the best option might be to cargo install it?

Yep, `cargo install just` is good enough.

**Comment by FollowTheProcess** [comment]
So this is all done with a few tweaks:

I've had to `cargo install just` as the `vscode` user as doing so under `root` causes a permission error when running any just recipe that touches `cargo` (presumably because the user is trying to modify/execute stuff that was created under `root`?):

image

Second, it seems that the default python in the devcontainer is 3.7.3 so the walrus does indeed fail!

image

image


I'm not totally sure how to introduce another version of python, I don't think it's possible to just include something like `FROM python3:slim-buster`?

The only other way I could think is building from source but this seems complex and lengthy for a devcontainer? Are you aware of another way?

For now I've just used the 3.7 style syntax to get it working and pushed up, I'll have a look into it in the meantime 👍🏻

**Comment by brettcannon** [comment]
It's fine to leave out the walrus operator; I'm too lazy to update the dev container just to support it right now. 😄

When you're ready for another review, feel free to refresh my review so it ends back up into my GitHub review queue.

**Comment by FollowTheProcess** [comment]
> It's fine to leave out the walrus operator; I'm too lazy to update the dev container just to support it right now. 😄

In that case I think it should be good to go!

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
Thanks for all your work on this, @FollowTheProcess !

**Comment by FollowTheProcess** [comment]
> Thanks for all your work on this, @FollowTheProcess !

Happy to help! `py` is my go to these days, thanks for your time helping and reviewing 🎉

**Pull request merged**